interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    /// if let (Some(hd), l2) = l2.split() {
+    ///     let tup = l2.split();
+    /// }
+    /// l1.push(5);
+    /// ```
+    pub fn split(self) -> (Option<T>, Self) {
+        if self.is_empty() {
+            return (None, self);
+        } else {
+            if let Ok(head) = Rc::try_unwrap(self.head.unwrap()) {
+                let tail_list = List { head: head.next }; //better: not cloning the tail
+                return (Some(head.elem), tail_list);
+            } else {
+                panic!("Cannot unwrap the head of this list. You probably tried Smooshing while a fork exists!")
+            }
+        }
+    }
+
+    /// Returns an Option of a pointer to the head of this list
+    pub fn head(&self) -> Option<&T> {
+        self.head.as_ref().map(|node| &node.elem)
+    }
+
+    pub fn iter(&self) -> Iter<'_, T> {
+        Iter {
+            next: self.head.as_deref(),
+        }
+    }
+}
+
+pub struct Iter<'a, T> {
+    next: Option<&'a Node<T>>,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.map(|node| {
+            self.next = node.next.as_deref();
+            &node.elem
+        })
+    }
+}
+
+///A Stack of HashMaps that supports scoping
+///
+///Invariants:
+///=> To Smoosh X levels down, no fork may still exist from any node amongst those
+/// X levels. So the stack must look like:
+/// ```text
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+/// ```
+///=> To merge, both branches must share a common fork point
+/// ```text
+///  _
+/// |_|
+///  |
+///  _     _
+/// |_|   |_|
+///  |     |
+///   \ _ /
+///    |*|
+///     |
+/// ```
+#[derive(Debug)]
+pub struct Smoosher<K: Eq + std::hash::Hash, V: Eq> {
+    head: HashMap<K, V>,       //mutable
+    tail: List<HashMap<K, V>>, //read-only
+}
+
+impl<K: Eq + std::hash::Hash, V: Eq> Smoosher<K, V> {
+    pub fn new() -> Smoosher<K, V> {
+        Smoosher {
+            head: HashMap::new(),
+            tail: List::new(),
+        }
+    }
+
+    /// If [self] and [other] share a fork point, returns a pair (depthA, depthB)
+    /// of the depth which the fork point can be found in [self] and [other], respectively.
+    /// NOTE: should be private, only public for testing!
+    pub fn shared_fork_point(&self, other: &Self) -> Option<(u64, u64)> {
+        //check head
+        if std::ptr::eq(&self.head, &other.head) {
+            return Some((0, 0));
+        } else {
+            //check tail
+            //these start at 1 b/c head was 0
+            let mut a_depth = 1;
+            let mut b_depth = 1;
+            //iterate and check all other nodes
+            for nd in self.tail.iter() {
+                b_depth = 1;
+                for nd_other in other.tail.iter() {
+                    if std::ptr::eq(nd, nd_other) {
+                        return Some((a_depth, b_depth));
+                    }
```

Comment:
You should be able to use `find` since you've defined an iterator. It's not spiritually different, though you may need to define equality over nodes as pointer equality.

Reply:
Find will returns the element, while we need the index of the element. We will try to use position.

Reply:
Ah right, you can use find_map over the .enumerate() of the iterator, though you may need to adjust the index accordingly. You can also leave it the same, should you so desire